### PR TITLE
Support command with RAFT

### PR DESCRIPTION
**Pull request by dangleptr**
Support sending command with RAFT.  
When encountering a command log,   leader will hold the logs after it until the command log committed. 

It is the foundation for the following features include "Transfer leader", "member change" etc. 

Subtask of #515

### CODE REVIEW

**Comment by dangleptr** [comment]
Jenkins go

File: src/kvstore/raftex/RaftPart.h [hunk 1]
```diff
@@ -123,6 +128,11 @@ class RaftPart : public std::enable_shared_from_this<RaftPart> {
      ***************************************************************/
     folly::Future<AppendLogResult> casAsync(std::string log);
 
+    /**
+     * Asynchronously send one command.
+     * */
+    folly::Future<AppendLogResult> commandAsync(std::string log);
```
Comment (sherman-the-tank) [inline src/kvstore/raftex/RaftPart.h]: Maybe rename it to ```sendCommandAsync()``` to make it consistent with other ```verbXxxxAsync()``` methods?

File: src/kvstore/raftex/RaftPart.h [hunk 2]
```diff
@@ -325,6 +335,14 @@ class RaftPart : public std::enable_shared_from_this<RaftPart> {
             return singlePromises_.back().getFuture();
         }
 
+        folly::Future<ValueType> getAndRollSharedFuture() {
```
Comment (sherman-the-tank) [inline src/kvstore/raftex/RaftPart.h]: Not sure why you need to use SharedPromise here. I think what you need is a SinglePromise, just like CAS does

**Comment by dangleptr** [comment]
> From what I understand here, it seems COMMAND is very similar to CAS. The only difference is at the commit stage. So I think we can treat COMMAND same as CAS until we are ready to commit

Yes,  COMMAND is similar to CAS, but not the same.  There are two differences:
 * Normal logs after CAS could be committed together.   In opposite,  Normal logs after COMMAND should be hold until the COMMAND committed,  but the logs **before** COMMAND could be committed together.  This is the purpose about changes in LogIterator. 
 * CAS maybe failed.  So we use SinglePromise for it.  But COMMAND not,  so it could shared one promise with the normal logs before it.  This is the purpose about changes in PromiseSet. 

@sherman-the-tank

File: src/kvstore/raftex/RaftPart.h [hunk 3]
```diff
@@ -39,10 +39,23 @@ enum class AppendLogResult {
     E_WAL_FAILURE = -6,
 };
 
+enum class LogType {
+    NORMAL      = 0x00,
+    CAS         = 0x01,
+    /**
+      COMMAND is similar to CAS, but not the same. There are two differences:
```
Comment (sherman-the-tank) [inline src/kvstore/raftex/RaftPart.h]: Well done 👍

File: src/kvstore/raftex/RaftPart.h [hunk 4]
```diff
@@ -326,6 +344,14 @@ class RaftPart : public std::enable_shared_from_this<RaftPart> {
             return singlePromises_.back().getFuture();
         }
 
+        folly::Future<ValueType> getAndRollSharedFuture() {
+            if (rollSharedPromise_) {
+                sharedPromises_.emplace_back();
+            }
+            rollSharedPromise_ = true;
```
Comment (critical27) [inline src/kvstore/raftex/RaftPart.h]: It really confuses me about rollSharedPromise_ at first :joy:.

**Pull request merged**